File: java/src/main/java/cloud/operon/sdk/SessionValidator.java
package cloud.operon.sdk;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.JsonNode;
import cloud.operon.sdk.auth.DecodedClaims;
import cloud.operon.sdk.internal.ApiErrorDecoder;
import cloud.operon.sdk.internal.Json;

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.time.Instant;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Session validation helper for PAT-based flows.
 */
public final class SessionValidator {

    private SessionValidator() {
    }

    public static SessionInfo validateSession(SessionValidationConfig cfg, String pat) throws OperonException {
        String normalizedPat = trimToNull(pat);
        if (normalizedPat == null) {
            throw new OperonException("pat is required");
        }

        String baseUrl = cfg == null ? null : trimToNull(cfg.baseUrl());
        if (baseUrl == null) {
            baseUrl = Config.DEFAULT_BASE_URL;
        }
        if (baseUrl.endsWith("/")) {
            baseUrl = baseUrl.substring(0, baseUrl.length() - 1);
        }

        HttpClient client = cfg == null ? null : cfg.httpClient();
        if (client == null) {
            client = HttpClient.newBuilder().connectTimeout(Duration.ofSeconds(30)).build();
        }

        HttpRequest request = HttpRequest.newBuilder()
            .uri(URI.create(baseUrl + "/v1/session/validate"))
            .header("Authorization", "Bearer " + normalizedPat)
            .header("Accept", "application/json")
            .GET()
            .build();

        HttpResponse<java.io.InputStream> response;
        try {
            response = client.send(request, HttpResponse.BodyHandlers.ofInputStream());
        } catch (InterruptedException ex) {
            Thread.currentThread().interrupt();
            throw new OperonException("perform validation request interrupted", ex);
        } catch (IOException ex) {
            throw new OperonException("perform validation request: " + ex.getMessage(), ex);
        }

        try (java.io.InputStream stream = response.body()) {
            if (response.statusCode() >= 400) {
                throw ApiErrorDecoder.decode(response.statusCode(), stream);
            }

            JsonNode payload = Json.mapper().readTree(stream);
            DecodedClaims claims = DecodedClaims.decode(normalizedPat);

            Instant expiresAt = null;
            if (claims.expiresAtUnix() > 0) {
                expiresAt = Instant.ofEpochSecond(claims.expiresAtUnix());
            }

            int expiresInSeconds = 0;
            if (expiresAt != null) {
                long delta = Duration.between(Instant.now(), expiresAt).getSeconds();
                expiresInSeconds = (int) Math.max(delta, 0L);
            }

            List<String> roles = new ArrayList<>();
            JsonNode rolesNode = payload.path("roles");
            if (rolesNode.isArray()) {
                rolesNode.forEach(item -> {
                    if (item.isTextual() && !item.asText().isBlank()) {
                        roles.add(item.asText());
                    }
                });
            }

            Map<String, Object> featureFlags = new HashMap<>();
            JsonNode flagsNode = payload.path("feature_flags");
            if (flagsNode.isObject()) {
                featureFlags = Json.mapper().convertValue(flagsNode, new TypeReference<Map<String, Object>>() {
                });
            }

            String clientId = firstNonEmpty(claims.clientId(), claims.authorizedParty());

            return new SessionInfo(
                payload.path("user_id").asText(""),
                payload.path("email").asText(""),
                payload.path("name").asText(""),
                payload.path("customer_id").asText(""),
                List.copyOf(roles),
                Map.copyOf(featureFlags),
                firstNonEmpty(claims.workstreamId()),
                firstNonEmpty(claims.workspaceId()),
                firstNonEmpty(claims.participantDid()),
                firstNonEmpty(claims.participantId()),
                clientId,
                firstNonEmpty(claims.sessionId()),
                expiresAt,
                expiresInSeconds
            );
        } catch (IOException ex) {
            throw new OperonException("decode validation response: " + ex.getMessage(), ex);
        }
    }

    private static String firstNonEmpty(String... values) {
        if (values == null) {
            return "";
        }
        for (String value : values) {
            String trimmed = trimToNull(value);
            if (trimmed != null) {
                return trimmed;
            }
        }
        return "";
    }

    private static String trimToNull(String value) {
        if (value == null) {
            return null;
        }
        String trimmed = value.trim();
        return trimmed.isEmpty() ? null : trimmed;
    }
}

File: dotnet/Operon.Sdk/SessionValidator.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Net.Http.Json;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Threading;
using System.Threading.Tasks;
using Operon.Sdk.Auth;
using Operon.Sdk.Errors;
using Operon.Sdk.Internal;
using Operon.Sdk.Models;

namespace Operon.Sdk;

/// <summary>
/// PAT session validation helper.
/// </summary>
public static class SessionValidator
{
    private static readonly JsonSerializerOptions SerializerOptions = new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
        PropertyNameCaseInsensitive = true,
        DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull
    };

    /// <summary>
    /// Validates PAT session metadata via the client API.
    /// </summary>
    public static async Task<SessionInfo> ValidateSessionAsync(
        SessionValidationConfig config,
        string pat,
        CancellationToken cancellationToken = default)
    {
        if (string.IsNullOrWhiteSpace(pat))
        {
            throw new ValidationException("pat is required");
        }

        var normalized = NormalizeConfig(config);
        try
        {
            using var request = new HttpRequestMessage(
                HttpMethod.Get,
                new Uri(normalized.BaseUri, "v1/session/validate"));
            request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
            request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", pat.Trim());

            using var response = await normalized.HttpClient.SendAsync(request, cancellationToken).ConfigureAwait(false);
            if (!response.IsSuccessStatusCode)
            {
                throw await SdkModelHelpers.DecodeApiErrorAsync(response).ConfigureAwait(false);
            }

            var payload = await response.Content.ReadFromJsonAsync<ValidationResponse>(SerializerOptions, cancellationToken)
                .ConfigureAwait(false)
                ?? new ValidationResponse();

            var claims = DecodedClaims.Decode(pat);
            var expiresAt = claims.ExpiresAtUnix > 0
                ? DateTimeOffset.FromUnixTimeSeconds(claims.ExpiresAtUnix)
                : (DateTimeOffset?)null;
            var expiresIn = expiresAt.HasValue
                ? Math.Max(0, (int)Math.Floor((expiresAt.Value - DateTimeOffset.UtcNow).TotalSeconds))
                : 0;

            var featureFlags = payload.FeatureFlags ?? new Dictionary<string, object?>();
            var roles = payload.Roles?.Where(role => !string.IsNullOrWhiteSpace(role)).Select(role => role.Trim()).ToArray()
                ?? Array.Empty<string>();

            return new SessionInfo
            {
                UserId = payload.UserId ?? string.Empty,
                Email = payload.Email ?? string.Empty,
                Name = payload.Name ?? string.Empty,
                CustomerId = payload.CustomerId ?? string.Empty,
                Roles = roles,
                FeatureFlags = featureFlags,
                WorkstreamId = claims.WorkstreamId ?? string.Empty,
                ChannelId = claims.WorkstreamId ?? string.Empty,
                WorkspaceId = claims.WorkspaceId ?? string.Empty,
                ParticipantDid = claims.ParticipantDid ?? string.Empty,
                ParticipantId = claims.ParticipantId ?? string.Empty,
                ClientId = SdkModelHelpers.Coalesce(claims.ClientId, claims.AuthorizedParty) ?? string.Empty,
                SessionId = claims.SessionId ?? string.Empty,
                ExpiresAt = expiresAt,
                ExpiresInSeconds = expiresIn
            };
        }
        finally
        {
            if (normalized.OwnsHttpClient)
            {
                normalized.HttpClient.Dispose();
            }
        }
    }

    private static (Uri BaseUri, HttpClient HttpClient, bool OwnsHttpClient) NormalizeConfig(SessionValidationConfig config)
    {
        ArgumentNullException.ThrowIfNull(config);

        var baseUri = (config.BaseUri ?? OperonConfig.DefaultBaseUri).EnsureTrailingSlash();
        if (!baseUri.IsAbsoluteUri)
        {
            throw new ValidationException("baseUri must be absolute");
        }

        if (config.HttpClient is null)
        {
            var client = new HttpClient
            {
                BaseAddress = baseUri,
                Timeout = TimeSpan.FromSeconds(30)
            };
            return (baseUri, client, true);
        }

        if (config.HttpClient.BaseAddress is null)
        {
            config.HttpClient.BaseAddress = baseUri;
        }

        return (baseUri, config.HttpClient, false);
    }

    private sealed record ValidationResponse
    {
        [JsonPropertyName("user_id")]
        public string? UserId { get; init; }

        [JsonPropertyName("email")]
        public string? Email { get; init; }

        [JsonPropertyName("name")]
        public string? Name { get; init; }

        [JsonPropertyName("customer_id")]
        public string? CustomerId { get; init; }

        [JsonPropertyName("roles")]
        public string[]? Roles { get; init; }

        [JsonPropertyName("feature_flags")]
        public Dictionary<string, object?>? FeatureFlags { get; init; }
    }
}
